src/model.ts

```diff
@@ -90,23 +111,47 @@ export abstract class Model<T extends Model<T>> {
       await data.remove(this.__snapshot)
     }
 
-    const labels: {[label in Labels]?: string} = {}
-    for (const key of keys) {
-      if (isSecondary(key.index)) {
-        labels[key.index.options.label] = key.query()
-      }
+    await data.set(primary, snakecase(prune(this), { deep: true }), this.secondaries(keys))
+
+    if (this.shadowKeys) {
+      const shadowKeys = this.shadowKeys()
+      const shadowPrimaries = shadowKeys.map((shadowkeys) =>
+        this.primary(shadowkeys)
+      )
+      await Promise.all([
+        ...this.__shadowSnapshots.map((shadowSnapshot, index) => {
+          if (
+            shadowSnapshot &&
+                  shadowPrimaries[index] !== shadowSnapshot
+          ) {
+            return data.remove(shadowSnapshot)
+          }
+        }),
+        ...shadowPrimaries.map((shadowPrimary, index) =>
+          data.set(
+            shadowPrimary,
+            snakecase(prune(this), { deep: true }),
+            this.secondaries(shadowKeys[index])
+          )
+        ),
+      ])
     }
-
-    await data.set(primary, snakecase(prune(this), { deep: true }), labels)
   }
 
   /**
    *
-   * Will delete the model from the database.
+   * Will delete the model and any shadows from the database.
    *
    */
   async delete() {
     await data.remove(this.primary())
+    if (this.shadowKeys) {
+      await Promise.all(
+        this.shadowKeys().map((shadowKeys) =>
```

Comment:
While this does not strictly violate the variable shadowing rule, I suspect it would also be more clearer if the inner parameter here was called `keySet` or something in those lines as well.